src/interrupts.rs

```diff
@@ -0,0 +1,326 @@
+                    "failed to VFIO_DEVICE_GET_IRQ_INFO for index {}. Errno: {}",
+                    index,
+                    std::io::Error::last_os_error()
+                )
+                .into());
+            }
+
+            if (irq_info.flags & VFIO_IRQ_INFO_EVENTFD) == 0 {
+                continue;
+            }
+
+            self.interrupt_type = index;
+            return Ok(());
+        }
+
+        self.interrupt_type = 0;
+        Ok(())
+    }
+}
+
+impl InterruptsQueue {
+    /// Add the `event_fd` file descriptor to epoll.
+    pub fn vfio_epoll_ctl(&mut self, event_fd: RawFd) -> Result<(), Box<dyn Error>> {
+        let mut event: Event = Event {
+            events: libc::EPOLLIN as u32,
+            data: event_fd as u64,
+        };
+
+        let epoll_fd: RawFd = unsafe { libc::epoll_create1(0) };
+        if epoll_fd == -1 {
+            return Err(format!(
+                "failed to epoll_create1. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+
+        if unsafe {
+            libc::epoll_ctl(
+                epoll_fd,
+                libc::EPOLL_CTL_ADD,
+                event_fd,
+                &mut event as *mut _ as *mut libc::epoll_event,
+            )
+        } == -1
+        {
+            return Err(format!(
+                "failed to epoll_ctl. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+
+        self.vfio_epoll_fd = epoll_fd;
+        Ok(())
+    }
+
+    /// Waits for events on the epoll instance referred to by the file descriptor `epoll_fd`.
+    ///
+    /// The memory area pointed to by events will contain the events that will be available for the caller.
+    /// The `timeout` argument specifies the minimum number of milliseconds that epoll_wait will block.
+    /// Specifying a `timeout` of -1 causes epoll_wait to block indefinitely, while specifying a
+    /// `timeout` equal to zero cause epoll_wait to return immediately, even if no events are available.
+    /// Returns the number of ready file descriptors.
+    pub fn vfio_epoll_wait(&self, timeout: i32) -> Result<usize, Box<dyn Error>> {
+        let mut events = [Event::default(); 10];
+        let rc: usize;
+
+        let status = unsafe {
+            libc::epoll_wait(
+                self.vfio_epoll_fd,
+                events.as_mut_ptr() as *mut libc::epoll_event,
+                events.len() as i32,
+                timeout,
+            )
+        };
+        if status == -1 {
+            return Err(format!(
+                "failed to epoll_wait. Errno: {}",
+                std::io::Error::last_os_error()
+            )
+            .into());
+        }
+        rc = status as usize;
+        if rc > 0 {
+            /* epoll_wait has at least one fd ready to read */
+            for event in events.iter().take(rc) {
+                let mut val: u16 = 0;
```

Comment:
[In the C code](https://github.com/emmericp/ixy/blob/master/src/libixy-vfio.c#L191), `val` is a `uint64_t`. What is correct?

Reply:
Does not matter ;) you specify the size it should read anyway so it ready either `u16` or a `u64`. We just do this to reset the interrupt in the eventfd file so we don't care about how much we read but just that we read something :)

Reply:
I changed it to `u64` so it is consistent with the C code :)